PML AST and Analysis

c-brenn opened the pull request:
This PR adds the framework for doing more detailed analysis on PML. Opening this for review and discussion.

Previously, the parser would look at the token stream generated by the lexer, and discard anything that wasn't a drug. Now we need to do more detailed analysis, so the parser turns the token stream into an Abstract Syntax Tree.

Each node in the AST has the structure: `{type, attributes, value}`

Where

- type is an atom identifying the type of node. e.g. `:process`
- attributes is a keyword list. e.g. `[line: 4, name: "foo"]`
- value is either a list of children or a primitive value. e.g.` [{..., ..., ...}, {..., ..., ...}]` or `"string"` or `"1234"`

For a larger example. The following PML:

```
process foo {
  task bar {
    action baz {
      tool { "pills" }
      script { "eat the pills" }
      agent { "patient" }
      requires {
        drug { "torasemide" }
      }
      provides { "a cured patient" }
    }
    action baz2 {
      tool { "pills" }
      script { "eat the pills" }
      agent { (intangible)(inscrutable) pml.wtf && ("foo" || 1 != 2) }
      requires {
        drug { "trandolapril" }
      }
      provides { "a cured patient" }
    }
  }
}
```

Becomes this AST:

```elixir
{:process, [line: 1, name: 'foo'],
  [{:task, [line: 2, name: 'bar'],
    [{:action, [line: 3, name: 'baz'],
      [{:tool, [line: 4], "pills"}, {:script, [line: 5], "eat the pills"},
       {:agent, [line: 6], "patient"},
       {:requires, [line: 7], {:drug, [line: 8], "torasemide"}},
       {:provides, [line: 10], "a cured patient"}]},
     {:action, [line: 12, name: 'baz2'],
      [{:tool, [line: 13], "pills"}, {:script, [line: 14], "eat the pills"},
       {:agent, [line: 15],
        {:&&, [line: 15],
         [{:., [line: 15],
           [[{:prefix, [], "intangible"}, {:prefix, [], "inscrutable"},
             {:ident, [], "pml"}], "wtf"]},
          {:parenthesised, [],
           {:||, [line: 15], ["foo", {:!=, [line: 15], ["1", "2"]}]}}]}},
       {:requires, [line: 16], {:drug, [line: 17], "trandolapril"}},
       {:provides, [line: 19], "a cured patient"}]}]}]}}
```

The `Analysis` module iterates over the AST performing analysis. This analysis can then be returned to the user. For now, it simply walks the AST and collects the drugs.

houli committed:
WIP PML AST LOL

c-brenn committed:
generalise AST helper functions [WIP][skip ci]

also adds a testing function to Parser bc typing in iex is hard.

c-brenn committed:
extend AST [WIP][skip ci]

* expressions are now represented in the AST `{func, attrs, [arg1, arg2]}`.
* removes the ident special case for `manual` and `executable` - these are now
  keywords.

c-brenn committed:
adds analysis module

* changes parser.yrl so all tuples are 3-tuples
* removes drug identification from parser.ex
* adds analysis module (very basic atm) - can identify drugs
* PmlController uses the analysis module after parse
* fixes up tests

c-brenn commented:
Heading out for the day so opening this for some 👀 .

The test functions are useful for debugging in `iex`. Will remove them once we're happy with the PR.

c-brenn commented:
I see now that the `[prefix: ..., prefix: ..., ident: ...]` construct does not conform to the `{type, attrs, value}` structure I mentioned above. This part of the AST represents the awful variable access that PML lets you do (see the spec).

For now this may not matter, the Analysis module seems to be able to handle it, but we probably want to change this at some point

houli committed:
More AST generation

Convert lists to binaries
Ident constructs
Number conversion
String expression constructs

houli commented:
What we really need to make sure this is all hunky dory, is a pretty printer that should give out, (close enough to, parens x && y vs (x && y) and number formatting 1.0 vs 1 etc.) the same output as the input given to the parser

c-brenn commented on panacea/src/pml_parser.yrl:
```diff
@@ -169,19 +169,33 @@ operator -> '>=' : '$1'.
     list_to_binary(CharList).
 
-construct({Type, Line}, Attributes, Value) ->
+number_expression({_, Line, Num}) ->
```
I don't think we need to bother parsing numbers because we never actually use them.

c-brenn commented:
agreed, we need a function from AST to PML. The less processing we do of un-interesting parts of the AST, the easier this function will be. That is, if we just store the matched characters for numbers, variables and expressions, we don't need to reconstruct them.

c-brenn committed:
make the value of expressions a string

We never need to worry about the contents/structure of an expression. So instead
of builing out ASTs for expressions, this change just store them as strings.

houli commented:
One thing about the latest commit is that you lose the quotes around string expressions

c-brenn commented:
Good spot. I think the only place we should strip quotes is for drugs.

c-brenn committed:
don't strip any quotes while parsing

the only time we want to strip quotes is when extracting drugs - so we're better
off doing it in the analysis module

c-brenn committed:
adds module for turning an AST into PML

c-brenn committed:
remove test functions

c-brenn commented:
As a summary of the changes:

# Expressions
 expressions are now stored as plain old strings. We never do anything with them, so there's no real need to build out AST nodes for them.

# Strings
 We used to strip the quotes off strings in the parser. i.e. `'"foo"'` became `'foo'`. This was so we could return the drugs as they came out of the parser. Because we keep around more than just drugs now, it doesn't make sense to strip quotes off every string.

Stripping the quotes off drugs that are being sent to the users is done during analysis instead.

# AST to PML
There is now a module for turning an AST into PML again `Panacea.Pml.Ast`. It returns the PML as an [IO-List](https://www.bignerdranch.com/blog/elixir-and-io-lists-part-1-building-output-efficiently/). These are much cheaper to build than strings and can be used in any IO (we're just going to be sending this PML to the user).

c-brenn committed:
sync up fixture files - athloi needs this inscrutable pml

22a reviewed (approved):
Incredible!
